src/core/extra/msi_credential.hpp

```diff
@@ -0,0 +1,113 @@
+#include <string>
+#include <vector>
+#include <mutex>
+#include <storage_credential.h>
+#include <curl/curl.h>
+#include <rapidjson/document.h>
+
+namespace azure {
+    namespace storage_lite {
+        std::size_t write_callback(char *in, size_t size, size_t nmemb, std::string *out) {
+            std::size_t total_size = size * nmemb;
+            if (total_size) {
+                out->append(in, total_size);
+                return total_size;
+            }
+            return 0;
+        }
+
+        class msi_credential : public storage_credential {
+        public:
+            AZURE_STORAGE_API msi_credential();
+
+            void sign_request(const storage_request_base &, http_base &, const storage_url &,
+                              const storage_headers &) const override;
+
+            void set_msi_url(const std::string &msi_url);
+
+            bool refresh_token(std::string &error) const;
+
+        private:
+            std::string m_msi_url;
+            mutable std::mutex m_token_mutex;
+            mutable time_t m_token_expires;
+            mutable std::string m_token;
+        };
+
+        const std::string DEFAULT_MSI_URL = "http://169.254.169.254/metadata/identity/oauth2/token";
+
+        AZURE_STORAGE_API msi_credential::msi_credential()
+                :m_msi_url(DEFAULT_MSI_URL), m_token_expires(0), m_token("") {
+        }
+
+        void msi_credential::sign_request(const storage_request_base &, http_base &h, const storage_url &,
+                                          const storage_headers &) const {
+            std::string error;
+            if (refresh_token(error)) {
+                std::string authorization("Bearer ");
+                authorization.append(m_token);
+                h.add_header(constants::header_authorization, authorization);
+            } else {
+                std::cerr << error << std::endl;
+            }
+        }
+
+        bool msi_credential::refresh_token(std::string &error) const {
+            std::lock_guard <std::mutex> lg(m_token_mutex);
```

Comment:
I don't have the full context for this lock, but it seems this line locks a very large piece of code, could the critical region be reduced?

Reply:
I think we don't want more than one `refresh_token` happens at the same time.  So, if there are simultaneous requests running, the first request will trigger `refresh_token`, and the rest of them will wait. After the first one returns successfully, the `expires_on` time will be updated and the rest request should return quickly because the next if statement will match.